File: Feedback/FeedbackSubview.m
//
//  FeedbackSubview.m
//  Feedback
//
//  Created by Dylan Gyesbreghs on 12/07/16.
//  Copyright © 2016 Dylan Gyesbreghs. All rights reserved.
//

#import "FeedbackSubview.h"

#import <PureLayout.h>

@interface FeedbackSubview()

@property (nonatomic, strong) UIImageView *feedbackImageView;
@property (nonatomic, strong) UILabel *feedbackTitleLabel;

@end

@implementation FeedbackSubview

-(instancetype)initWithDelegate:(id <FeedbackSubviewDelegate>)delegate {
    if (self = [super init]) {
        _delegate = delegate;
        [self setupView];
        [self setupConstraint];
        [self setupAction];
    }
    return self;
}

- (void)setupView {
    _feedbackImageView = [UIImageView newAutoLayoutView];
    [_feedbackImageView.layer setCornerRadius:([_delegate sizeForSubview].width / 2)];
    [_feedbackImageView.layer setMasksToBounds:true];
    [_feedbackImageView setBackgroundColor:[UIColor greenColor]];
    [self addSubview:_feedbackImageView];
    
    _feedbackTitleLabel = [UILabel newAutoLayoutView];
    [self addSubview:_feedbackTitleLabel];
}

- (void)setupConstraint {
    if (_delegate == nil) {
        return;
    }
    
    [_feedbackImageView autoSetDimension:ALDimensionHeight toSize:[_delegate sizeForSubview].height];
    [_feedbackImageView autoSetDimension:ALDimensionWidth toSize:[_delegate sizeForSubview].width];
    [_feedbackImageView autoAlignAxisToSuperviewAxis:ALAxisHorizontal];
    [_feedbackImageView autoAlignAxisToSuperviewAxis:ALAxisVertical];
    
    [_feedbackTitleLabel autoAlignAxis:ALAxisVertical toSameAxisOfView:_feedbackImageView withOffset:12];
}

- (void)setupAction {
    
}

@end

File: Feedback/FeedbackSubview.swift
//
//  FeedbackSubview.swift
//  Feedback
//
//  Created by Dylan Gyesbreghs on 12/07/16.
//  Copyright © 2016 Dylan Gyesbreghs. All rights reserved.
//

import UIKit
import PureLayout

enum FeedbackSubviewType {
    case Bad
    case Okay
    case Good
    case Great
}

protocol FeedbackSubViewDatasource {
    func title(forType type : FeedbackSubviewType) -> String
    func image(forType type : FeedbackSubviewType) -> UIImage
}

class FeedbackSubview: UIView {
    var datasource : FeedbackSubViewDatasource?
    
    var itemLabel = UILabel.newAutoLayoutView()
    var logoImageView = UIImageView.newAutoLayoutView()
    
    private var heightConstraint : NSLayoutConstraint?
    private var widthConstraint : NSLayoutConstraint?
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        setupView()
        setupConstraint()
    }
    
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
    }
    
    func setupView() {
        addSubview(itemLabel)
        addSubview(logoImageView)
        
        logoImageView.layer.cornerRadius = 49 / 2
        logoImageView.layer.masksToBounds = true
        logoImageView.backgroundColor = UIColor.cyanColor()
    }
    
    func setupView(withType type : FeedbackSubviewType) {
        if let datasource = datasource {
            itemLabel.text = datasource.title(forType: type)
            logoImageView.image = datasource.image(forType: type)
        }
    }
    
    func setupConstraint() {
        heightConstraint = logoImageView.autoSetDimension(.Height, toSize: 49)
        widthConstraint = logoImageView.autoSetDimension(.Width, toSize: 49)
        logoImageView.autoAlignAxisToSuperviewAxis(.Horizontal)
        logoImageView.autoAlignAxisToSuperviewAxis(.Vertical)
        
        itemLabel.autoAlignAxis(.Vertical, toSameAxisOfView: logoImageView)
        itemLabel.autoPinEdge(.Top, toEdge: .Bottom, ofView: logoImageView, withOffset: 12)
    }
}
